### PR TITLE
[KLU] Initial pass

### PR DESCRIPTION
**Pull request by rayegun**
Planning to finish this up tomorrow. I've done some manual testing, but still need to add them.

Biggest question is whether the overall design seems reasonable so far. It's pretty simple.

### CODE REVIEW

**Comment by rayegun** [comment]
I still have a few things to add, more defensive checks in some places, a few more tests, and need to push my docs. However this is mostly ready for review.

In particular I'd like feedback on names and correctness of `ldiv!` and `\` interfaces, and whether I should wrap the following functions:

- `klu_flops`: determines # of flops performed in numeric factorzation
- `klu_condest`: Computes a reasonably accurate estimate of the 1-norm condition number
- `klu_rgrowth` : compute the reciprocal pivot growth
These are all fairly easily accessible through the Clang wrapper, and also through the `common` field of each `KLUFactorization` struct. There are also some other options in that field which I may add convenience wrappers for if that seems desirable.

**Comment by ViralBShah** [comment]
On `ldiv!` and `\`, it would be great to have @KristofferC and/or @dkarrasch take a look at it.

File: src/klu.jl [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+            else
+                K._numeric = num
+            end
+            return K
+        end
+    end
+end
+
+"""
+klu_factor!(K::KLUFactorization)
```
Comment (dkarrasch) [inline src/klu.jl]:
```suggestion
    klu_factor!(K::KLUFactorization)
```

File: src/klu.jl [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+
+Factor `K` into components `K.L`, `K.U`, and `K.F`.
+This function will perform both the symbolic and numeric steps of factoriation on an
+    existing `KLUFactorization` instance.
+
+    The `K.common` struct can be used to modify certain options and parameters such as:
+
```
Comment (dkarrasch) [inline src/klu.jl]:
```suggestion
existing `KLUFactorization` instance.

The `K.common` struct can be used to modify certain options and parameters such as:

```
Alos, this looks like unfinished.

File: src/klu.jl [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+- [`\\`](@ref)
+
+!!! note
+    `klu(A::SparseMatrixCSC)` uses the KLU library that is part of
+    SuiteSparse. As this library only supports sparse matrices with [`Float64`](@ref) or
+    `ComplexF64` elements, `lu` converts `A` into a copy that is of type
+    `SparseMatrixCSC{Float64}` or `SparseMatrixCSC{ComplexF64}` as appropriate.
+"""
+function klu(n, colptr::Vector{Ti}, rowval::Vector{Ti}, nzval::Vector{Tv}) where {Ti<:KLUITypes, Tv<:AbstractFloat}
+    if nzval != Float64
```
Comment (dkarrasch) [inline src/klu.jl]: I guess this is something like `Tv !== Float64`, i.e., `eltype(nzval) !== Float64`? Similar below.

File: src/klu.jl [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+
+for Tv ∈ KLUValueTypes, Ti ∈ KLUIndexTypes
+    tsolve = _klu_name("tsolve", Tv, Ti)
+    if Tv === :ComplexF64
+        call = :($tsolve(klu._symbolic, klu._numeric, size(B, 1), size(B, 2), B, conj, Ref(klu.common)))
+    else
+        call = :($tsolve(klu._symbolic, klu._numeric, size(B, 1), size(B, 2), B, Ref(klu.common)))
+    end
+    @eval begin
+        function solve!(klu::LinearAlgebra.Adjoint{$Tv, KLUFactorization{$Tv, $Ti}}, B::StridedVecOrMat{$Tv})
```
Comment (dkarrasch) [inline src/klu.jl]: Do you really need the qualifier? Similarly below with `Transpose`.

**Comment by rayegun** [comment]
Also not sure if `klu` is the right name for the primary function. It can't really be `lu` though. `luf` (`L * U + F`) might be better than `klu`?

**Comment by ViralBShah** [comment]
I think klu is fine for now, and keep everything as unexported APIs. One thought is, can lu do something like sort which takes different algorithms as it's input.

File: src/klu.jl (hunk 4 above)
Comment (rayegun) [reply src/klu.jl]: You mean drop the `LinearAlgebra.`?

Comment (dkarrasch) [reply src/klu.jl]: I think `Adjoint` is exported by `LinearAlgebra`, so there should be no issue.

**Comment by rayegun** [comment]
What is needed to get this past the finish line? I'm going to bulk up the tests and check the docs again but are there any other issues?

**Comment by ViralBShah** [comment]
Looks like we should be able to merge when all green - since all the comments are addressed. @dkarrasch @KristofferC Do you guys have anything further to add?

**Comment by rayegun** [comment]
I have absolutely no idea why the threading test is suddenly failing on Windows? It wasn't 1 commit ago, and I've never touched that test. I have also locally tested on x64 Windows with no problems...

Additionally I'm not sure I can realistically get a much higher coverage. Most of those lines are errors like OOM, or name generating functions not involved in actual usage.

**Comment by ViralBShah** [comment]
Can you re-run the CI to see if we can get all green?

**Comment by ViralBShah** [comment]
cc @DilumAluthge - who pointed out that we are seeing test failures in master in suitesparse as well.

**Comment by Gnimuc** [comment]
> OpenBLAS: malloc failed in gemm_driver

Looks like something weird just happened.

```
Test Summary: | Pass  Total
KLU Wrappers  |  129    129
OpenBLAS: malloc failed in gemm_driver
Test Summary:              | Pass  Total
threaded SuiteSparse tests |    1      1
ERROR: LoadError: SuiteSparse threads test failed with nthreads == 4
```

I cannot reproduce JuliaSparse/SuiteSparse.jl#45 on my local machine. Any thoughts? 

cc @DilumAluthge

**Comment by ViralBShah** [comment]
We need to rebuild the openblas on master.

**Comment by ViralBShah** [comment]
openblas has been rebuilt on master and new nightlies should be out that might fix the openblas gemm_driver issue.

**Comment by ViralBShah** [comment]
Ok - so I am guessing some updates on julia master resolved this one.

**Comment by ViralBShah** [comment]
@Wimmerer Please merge if this is good to go.

**Review by ViralBShah — APPROVED** [review]

**Comment by ViralBShah** [comment]
Actually - let's hold until JuliaSparse/SparseArrays.jl#147 is resolved.

**Comment by ViralBShah** [comment]
@Wimmerer Let's prepare this for merging.

**Comment by rayegun** [comment]
Ok things seem ready from my pov. I know you gave go ahead to merge previously but just wanted to wait for ok here.

**Comment by ViralBShah** [comment]
As I review the code I am thinking it will be easier to be a separate package - especially since we are exporting APIs - and are not otherwise doing type piracy and such. That way it can be updated with less effort. This doesn't seem to need any type piracy at the moment. Perhaps having a KLU.jl package may be better - while depending on the suitesparse libraries that are currently provided by Julia?

**Comment by rayegun** [comment]
That's absolutely fine by me, it also means Chris will get this much sooner.

I'll try to submit it later today.

**Comment by ViralBShah** [comment]
Yes, otherwise we have to wait months. I am closing this one. Thank you.

**Comment by ViralBShah** [comment]
Also KLU is [LGPL](https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/KLU/Doc/lesser.txt) - which is another good reason to keep it separate.

**Pull request closed without merging**